teps/0090-matrix.md

```diff
@@ -596,17 +627,546 @@ For example:
+```
+
+##### Substituting Array Parameters in the Tasks
+
+To substitute `Parameters` of type `Array` in `Tasks`, we would need
+`Parameters` of type `Arrays of Arrays` in the `Matrix` in `PipelineTasks`. 
+However, we currently support `Parameters` of type `String` and `Arrays` only. 
+
+For example, taking the [*gcloud* `Task`][gcloud-task] in the Tekton Catalog,
+which declares a *ARGS* array `Parameter`. Say we want to execute it thrice 
+to check authorization, deploy to Cloud Run, and create a GCE instance. And
+we want to leverage the `Matrix` to do all of that in one `PipelineTask`. This
+is the specification that we will need:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: gcloud-pipeline
+spec:
+  serviceAccountName: workload-identity-sa
+  tasks:
+  ...
+  - name: setup
+    taskRef:
+      name: gcloud
+    matrix:
+    - name: ARGS
+      value: [
+        ['auth', 'list'],
+        ['run', 'deploy', 'my-service', '--image=gcr.io/my-project/my-image', '--platform=PLATFORM', '--region=REGION'],
+        ['compute', 'instances', 'create', 'my-instance', '--zone=ZONE']
+      ]
+```
+
+As shown above, we would need `Parameters` of type `Arrays of Arrays` to 
+substituting `Arrays` in `Tasks` through the `Matrix`.
+
+In [TEP-0075:Object Parameter and Results][tep-0075], we are exploring supporting 
+more data structures in `Parameters` through [JSON object schema syntax][json-schema]. 
+TEP-0075 may provide `Parameters` of type `Arrays of Arrays`, then we can discuss 
+supporting it in `Matrix`.
+
+Moreover, [use cases](#use-cases) we are solving for in this TEP don't need this feature.
+We plan to provide minimum features need to meet the user cases, per the [simplicity][simplicity] 
+design principle. However, this remains an option we can explore later.
+
+### Variables in the Matrix
+
+Similarly to the `Parameters` in the `params` field, the `Parameters` in the 
+`matrix` field will accept variables that will be substituted, including:
+* `Results`: which are passed in from previous `PipelineTasks`.
+* `Execution Status`: which available in `Finally Tasks` only.
+* `Context Variables`: include `Task`, `Pipeline`, `TaskRun` and `PipelineRun` names.
+
+### Results from Matrixed PipelineTasks
+
+Today, each `PipelineTask` is executed in a single `TaskRun` or `Run` which produces 
+the `Results` specified in the underlying `Task` or `Custom Task`. The `Results` are 
+accessible through variable - `$(tasks.<pipelinetask-name>.results.<result-name>)`. 
+
+With the `Matrix`, the `PipelineTask` will be executed in several `TaskRuns` or `Runs`,
+each of which will produce its own `Results`. Therefore, we need to adjust the variables
+for `Results` when a `Matrix` is present to accommodate all the `Results` produced. 
+We propose suffixing the `PipelineTask` name in the `Result` variable with the count 
+of the matrix execution: `$(tasks.<pipelinetask-name>-<count>.results.<result-name>)`.
```

Comment:
how about referencing the results of the task with `$(tasks.<pipelinetask-name>-<matrixvalue[count]>.results.<result-name>)`

Reply:
moreover, how are we going to handle the workspaces if the task is writing something to it and we plan to run it parallelly?

Reply:
addressed this in the proposal: https://github.com/jerop/community/blob/matrix-proposal/teps/0090-matrix.md#workspaces